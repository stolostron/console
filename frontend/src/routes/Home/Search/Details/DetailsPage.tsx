/* Copyright Contributors to the Open Cluster Management project */
// Copyright (c) 2021 Red Hat, Inc.

import { useEffect, useMemo, useState } from 'react'
import { Link, Route, Switch, useLocation } from 'react-router-dom'
import { Pages, usePageVisitMetricHandler } from '../../../../hooks/console-metrics'
import { useTranslation } from '../../../../lib/acm-i18next'
import { NavigationPath } from '../../../../NavigationPath'
import { IResource } from '../../../../resources'
import { fireManagedClusterView } from '../../../../resources/managedclusterview'
import { getResource } from '../../../../resources/utils/resource-request'
import { AcmPage, AcmPageHeader, AcmSecondaryNav, AcmSecondaryNavItem } from '../../../../ui-components'
import DetailsOverviewPage from './DetailsOverviewPage'
import LogsPage from './LogsPage'
import RelatedResourceDetailsTab from './RelatedResourceDetailsTab'
import YAMLPage from './YAMLPage'

export function getResourceParams() {
  let cluster = '',
    kind = '',
    apiversion = '',
    namespace = '',
    name = ''
  const urlParams = decodeURIComponent(window.location.search).replace('?', '').split('&')
  urlParams.forEach((param) => {
    const paramKey = param.split('=')[0]
    const paramValue = param.split('=')[1]
    switch (paramKey) {
      case 'cluster':
        cluster = paramValue
        break
      case 'kind':
        kind = paramValue
        break
      case 'apiversion':
        apiversion = paramValue
        break
      case 'namespace':
        namespace = paramValue
        break
      case 'name':
        name = paramValue
        break
    }
  })
  return { cluster, kind, apiversion, namespace, name }
}

export default function DetailsPage() {
  usePageVisitMetricHandler(Pages.searchDetails)
  const { t } = useTranslation()
  const [resource, setResource] = useState<any>(undefined)
  const [containers, setContainers] = useState<string[]>()
  const [resourceVersion, setResourceVersion] = useState<string>('')
  const [resourceError, setResourceError] = useState('')
  const { cluster, kind, apiversion, namespace, name } = getResourceParams()

  useEffect(() => {
    if (resourceVersion !== resource?.metadata.resourceVersion || name !== resource?.metadata.name) {
      /* istanbul ignore else */
      if (cluster === 'local-cluster') {
        getResource<IResource>({
          apiVersion: apiversion,
          kind,
          metadata: { namespace, name },
        })
          .promise.then((response) => {
            setResource(response)
            setResourceVersion(response?.metadata?.resourceVersion ?? '')
          })
          .catch((err) => {
            console.error('Error getting resource: ', err)
            setResourceError(err.message)
          })
      } else {
        fireManagedClusterView(cluster, kind, apiversion, name, namespace)
          .then((viewResponse) => {
            if (viewResponse?.message) {
              setResourceError(viewResponse.message)
            } else {
              setResource(viewResponse?.result)
              setResourceVersion(viewResponse?.result?.metadata.resourceVersion ?? '')
            }
          })
          .catch((err) => {
            console.error('Error getting resource: ', err)
            setResourceError(err)
          })
      }
    }
  }, [
    cluster,
    kind,
    apiversion,
    name,
    namespace,
    resourceVersion,
    resource?.metadata.resourceVersion,
    resource?.metadata.name,
  ])

  useEffect(() => {
    setContainers((resource && resource.spec?.containers?.map((container: any) => container.name)) ?? [])
  }, [resource])

  const location: {
    pathname: string
    state: {
      from: string
      search: string
      fromSearch: string
    }
  } = useLocation()

  const breadcrumbLink: string = useMemo(() => {
    const prevLocState = location.state
    if (prevLocState && prevLocState.from === NavigationPath.search) {
      // If we came to resources page from search - return to search with previous search filters
      const previousSearchState = location.state.fromSearch ?? /* istanbul ignore next */ ''
      return `${NavigationPath.search}${previousSearchState}`
    } else {
      /* istanbul ignore next */
      return NavigationPath.search
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  return (
    <AcmPage
      header={
        <AcmPageHeader
          title={name}
          breadcrumb={[
            {
              text: t('Search'),
              to: breadcrumbLink,
            },
            {
              text: t('{{kind}} details', { kind }),
              to: '',
            },
          ]}
          navigation={
            <AcmSecondaryNav>
              <AcmSecondaryNavItem isActive={location.pathname === NavigationPath.resources}>
                <Link to={`${NavigationPath.resources}${window.location.search}`}>{t('Details')}</Link>
              </AcmSecondaryNavItem>
              <AcmSecondaryNavItem isActive={location.pathname === NavigationPath.resourceYAML}>
                <Link to={`${NavigationPath.resourceYAML}${window.location.search}`}>{t('YAML')}</Link>
              </AcmSecondaryNavItem>
              <AcmSecondaryNavItem isActive={location.pathname === NavigationPath.resourceRelated}>
                <Link to={`${NavigationPath.resourceRelated}${window.location.search}`}>{t('Related resources')}</Link>
              </AcmSecondaryNavItem>
              {(kind.toLowerCase() === 'pod' || kind.toLowerCase() === 'pods') && (
                <AcmSecondaryNavItem isActive={location.pathname === NavigationPath.resourceLogs}>
                  <Link to={`${NavigationPath.resourceLogs}${window.location.search}`}>{t('Logs')}</Link>
                </AcmSecondaryNavItem>
              )}
            </AcmSecondaryNav>
          }
        />
      }
    >
      <Switch>
        <Route exact path={NavigationPath.resources}>
          <DetailsOverviewPage
            cluster={cluster}
            loading={!resource && resourceError === ''}
            error={resourceError}
            resource={resource}
            name={name}
          />
        </Route>
        <Route exact path={NavigationPath.resourceYAML}>
          <YAMLPage
            resource={resource}
            loading={!resource && resourceError === ''}
            error={resourceError}
            name={name}
            namespace={namespace}
            cluster={cluster}
            kind={kind}
            apiversion={apiversion}
            setResourceVersion={setResourceVersion}
          />
        </Route>
        <Route exact path={NavigationPath.resourceRelated}>
          <RelatedResourceDetailsTab
            cluster={cluster}
            resource={resource}
            resourceLoading={!resource && resourceError === ''}
          />
        </Route>
        {(kind.toLowerCase() === 'pod' || kind.toLowerCase() === 'pods') && containers && (
          <Route path={NavigationPath.resourceLogs}>
            <LogsPage
              resource={resource}
              resourceError={resourceError}
              containers={containers}
              cluster={cluster}
              namespace={namespace}
              name={name}
            />
          </Route>
        )}
      </Switch>
    </AcmPage>
  )
}
