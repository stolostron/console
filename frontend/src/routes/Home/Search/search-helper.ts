/* Copyright Contributors to the Open Cluster Management project */
// Copyright (c) 2021 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project

import { DropdownSuggestionsProps } from './components/Searchbar'

const operators = ['=', '<', '>', '<=', '>=', '!=', '!']
const dateValues = ['hour', 'day', 'week', 'month', 'year']

export function formatSearchbarSuggestions(
    data: string[],
    suggestionKind: 'label' | 'filter' | 'value',
    searchQuery: string
) {
    let valuesToRemoveFromSuggestions: string[] = []
    let suggestions: DropdownSuggestionsProps[] = []
    const labelTag = {
        id: 'id-suggestions-label',
        key: 'key-suggestions-label',
        name: 'Filters',
        kind: suggestionKind,
        disabled: true,
    }
    if (suggestionKind === 'value') {
        // Get a list of duplicate values to remove from suggestions dropdown
        const searchTokens = searchQuery.split(' ')
        const searchCompleteFilter = searchTokens[searchTokens.length - 1].replace(':', '')
        labelTag.name = `${searchCompleteFilter} values`
        const query = convertStringToQuery(searchQuery)
        query.filters.forEach((filter) => {
            if (filter.property === searchCompleteFilter) {
                valuesToRemoveFromSuggestions = filter.values.filter((value) => data.indexOf(value) > 0)
            }
        })
        if (data[0] === 'isNumber') {
            if (operators.some((operator: string) => searchQuery.endsWith(operator))) {
                // operator is already chosen
                const numberRange =
                    data.length > 2
                        ? `Min: ${parseInt(data[1], 10)} - Max: ${parseInt(data[2], 10)}`
                        : `Min: ${parseInt(data[1], 10)} - Max: ${parseInt(data[1], 10)}`

                return [
                    labelTag,
                    {
                        id: 'id-values-range',
                        key: 'key-values-range',
                        name: numberRange,
                        kind: suggestionKind,
                        disabled: true,
                    },
                ]
            }
            suggestions = operators.map((op) => {
                return {
                    id: `id-${op}`,
                    key: `key-${op}`,
                    name: op,
                    kind: suggestionKind,
                }
            })
            suggestions.unshift({
                id: 'id-operator-label',
                name: 'Operators',
                kind: 'label',
                disabled: true,
            })
            return suggestions
        } else if (data[0] === 'isDate') {
            suggestions = dateValues.map((date) => {
                return {
                    id: `id-date-${date}`,
                    key: `key-date-${date}`,
                    name: date,
                    kind: suggestionKind,
                }
            })
            suggestions.unshift({
                id: 'id-filter-label',
                name: `${searchCompleteFilter} within the last:`,
                kind: 'label',
                disabled: true,
            })
            return suggestions
        }
    }

    suggestions = data
        .filter((suggestion) => {
            return valuesToRemoveFromSuggestions.indexOf(suggestion) === -1
        })
        .map((field) => {
            return {
                id: `id-${field}`,
                key: `key-${field}`,
                name: field,
                kind: suggestionKind,
            }
        })
    suggestions.unshift(labelTag)
    return suggestions
}

export const convertStringToQuery = (searchText: string) => {
    const searchTokens = searchText.split(' ')
    const keywords = searchTokens.filter((token) => token !== '' && token.indexOf(':') < 0)
    const filters = searchTokens
        .filter((token) => token.indexOf(':') >= 0)
        .map((f) => {
            const splitIdx = f.indexOf(':')
            const property = f.substring(0, splitIdx)
            const values = f.substring(splitIdx + 1)
            return { property, values: values.split(',') }
        })
        .filter((f) => ['', '=', '<', '>', '<=', '>=', '!=', '!'].findIndex((op) => op === f.values[0]) === -1)
    return {
        keywords,
        filters,
        // Limit results to  for better performance - we show a info alert to user asking them to refine search.
        limit: 1000,
    }
}

export const getSearchCompleteString = (searchQuery: string) => {
    const queryTags = searchQuery.split(' ')
    if (queryTags[queryTags.length - 1].endsWith(':')) {
        return queryTags[queryTags.length - 1].replace(':', '')
    } else if (
        operators.some(
            (op) => queryTags[queryTags.length - 1].substring(queryTags[queryTags.length - 1].length - op.length) === op
        )
    ) {
        const operator = operators.filter(
            (op) => queryTags[queryTags.length - 1].substring(queryTags[queryTags.length - 1].length - op.length) === op
        )
        return queryTags[queryTags.length - 1].replace(':', '').replace(operator[operator.length - 1], '')
    }
    return ''
}
