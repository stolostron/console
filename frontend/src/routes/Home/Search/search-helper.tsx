/* Copyright Contributors to the Open Cluster Management project */
// Copyright (c) 2021 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project

import { TFunction } from 'react-i18next'
import { Link } from 'react-router-dom-v5-compat'
import { NavigationPath } from '../../../NavigationPath'
import { SearchAlertProps } from './components/SearchAlertGroup'
import { DropdownSuggestionsProps } from './components/Searchbar'

export const federatedErrorText = 'error sending federated request'
export const operators = ['<=', '>=', '!=', '!', '=', '<', '>']
const dateValues = ['hour', 'day', 'week', 'month', 'year']

export function formatSearchbarSuggestions(
  data: string[],
  suggestionKind: 'label' | 'filter' | 'value',
  searchQuery: string,
  limit: number,
  isLoading: boolean,
  t: TFunction
) {
  const valuesToRemoveFromSuggestions: string[] = []
  let suggestions: DropdownSuggestionsProps[] = []
  const labelTag = {
    id: 'id-suggestions-label',
    name: t('Filters'),
    kind: suggestionKind,
    disabled: true,
  }
  if (suggestionKind === 'value') {
    // Get a list of duplicate values to remove from suggestions dropdown
    const searchTokens = searchQuery.split(' ')
    const searchCompleteFilter = searchTokens[searchTokens.length - 1].substring(
      0,
      searchTokens[searchTokens.length - 1].indexOf(':')
    )
    labelTag.name = t('{{0}} values', [searchCompleteFilter])
    const query = convertStringToQuery(searchQuery, limit)
    query.filters.forEach((filter) => {
      valuesToRemoveFromSuggestions.push(...(filter.property === searchCompleteFilter ? filter.values : []))
    })
    if (data[0] === 'isNumber') {
      if (operators.some((operator: string) => searchQuery.endsWith(operator))) {
        // operator is already chosen
        const numberRange =
          data.length > 2
            ? `Min: ${parseInt(data[1], 10)} - Max: ${parseInt(data[2], 10)}`
            : `Min: ${parseInt(data[1], 10)} - Max: ${parseInt(data[1], 10)}`

        return [
          labelTag,
          {
            id: 'id-values-range',
            name: numberRange,
            kind: suggestionKind,
            disabled: true,
          },
        ]
      }
      suggestions = operators.map((op) => {
        return {
          id: `id-${op}`,
          name: op,
          kind: suggestionKind,
        }
      })
      suggestions.unshift({
        id: 'id-operator-label',
        name: t('Operators'),
        kind: 'label',
        disabled: true,
      })
      return suggestions
    } else if (data[0] === 'isDate') {
      suggestions = dateValues.map((date) => {
        return {
          id: `id-date-${date}`,
          name: date,
          kind: suggestionKind,
        }
      })
      suggestions.unshift({
        id: 'id-filter-label',
        name: t('{{0}} within the last:', [searchCompleteFilter]),
        kind: 'label',
        disabled: true,
      })
      return suggestions
    }
  }

  suggestions = data
    .filter((suggestion) => {
      return valuesToRemoveFromSuggestions.indexOf(suggestion) === -1
    })
    .map((field) => {
      return {
        id: `id-${field}`,
        name: field,
        kind: suggestionKind,
      }
    })
  suggestions.unshift(labelTag)
  isLoading &&
    suggestions.push({
      id: 'id-loading',
      name: suggestionKind === 'filter' ? t('Loading...') : t('Loading more...'),
      kind: 'label',
      disabled: true,
    })
  return suggestions
}

export const convertStringToQuery = (searchText: string, queryResultLimit: number) => {
  const searchTokens = searchText.split(' ')
  const keywords = searchTokens.filter((token) => token !== '' && token.indexOf(':') < 0)
  const filters = searchTokens
    .filter((token) => token.indexOf(':') >= 0)
    .map((f) => {
      const splitIdx = f.indexOf(':')
      const property = f.substring(0, splitIdx)
      const values = f.substring(splitIdx + 1)
      return { property, values: values.split(',') }
    })

  return {
    keywords,
    filters,
    limit: queryResultLimit,
  }
}

export const getSearchCompleteString = (searchQuery: string) => {
  const queryTags = searchQuery.split(' ')
  if (queryTags[queryTags.length - 1].endsWith(':')) {
    return queryTags[queryTags.length - 1].replace(':', '')
  } else if (
    operators.some(
      (op) => queryTags[queryTags.length - 1].substring(queryTags[queryTags.length - 1].length - op.length) === op
    )
  ) {
    const operator = operators.filter(
      (op) => queryTags[queryTags.length - 1].substring(queryTags[queryTags.length - 1].length - op.length) === op
    )
    return queryTags[queryTags.length - 1].replace(':', '').replace(operator[0], '')
  }
  return ''
}

export function setFederatedErrorAlert(
  loading: boolean,
  error: any,
  data: any,
  alerts: SearchAlertProps[],
  addSearchAlert: (alert: SearchAlertProps) => void,
  removeSearchAlert: (key: string) => void,
  t: any
) {
  const federatedWarningKey = 'federated-error'
  if (!loading && error && error?.graphQLErrors.find((error: any) => error?.includes(federatedErrorText))) {
    !alerts.find((alert) => alert.key === federatedWarningKey) &&
      addSearchAlert({
        key: federatedWarningKey,
        variant: 'warning',
        title: t('One or more managed hubs are not reporting'),
        children: (
          <div style={{ display: 'flex', flexDirection: 'column' }}>
            {t(
              'Resources from these managed hubs and managed clusters will be missing in search results. Resolve cluster issues to see all cluster resources.'
            )}
            <Link style={{ paddingTop: '0.25rem' }} to={NavigationPath.clusters}>
              {t('View clusters')}
            </Link>
          </div>
        ),
      })
  } else if (
    !loading &&
    !error?.graphQLErrors.find((error: any) => error?.includes(federatedErrorText)) &&
    data &&
    alerts.find((alert) => alert.key === federatedWarningKey)
  ) {
    removeSearchAlert(federatedWarningKey)
  }
}
