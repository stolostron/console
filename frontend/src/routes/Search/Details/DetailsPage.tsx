/* Copyright Contributors to the Open Cluster Management project */
// Copyright (c) 2021 Red Hat, Inc.

import { Divider, Dropdown, DropdownItem, DropdownToggle } from '@patternfly/react-core'
import { Dispatch, SetStateAction, useContext, useEffect, useMemo, useState } from 'react'
import { Link, Outlet, useLocation, useNavigate, useOutletContext } from 'react-router-dom-v5-compat'
import { Pages, usePageVisitMetricHandler } from '../../../hooks/console-metrics'
import { useTranslation } from '../../../lib/acm-i18next'
import { NavigationPath } from '../../../NavigationPath'
import { IResource } from '../../../resources'
import { fireManagedClusterView } from '../../../resources/managedclusterview'
import { getResource } from '../../../resources/utils/resource-request'
import { useRecoilValue, useSharedAtoms } from '../../../shared-recoil'
import { AcmPage, AcmPageHeader, AcmSecondaryNav, AcmSecondaryNavItem, AcmToastContext } from '../../../ui-components'
import { handleVMActions } from '../SearchResults/utils'
import { DeleteResourceModal } from './DeleteResourceModal'

export type SearchDetailsContext = {
  cluster: string
  name: string
  namespace: string
  apiversion: string
  kind: string
  resource: any
  resourceLoading: boolean
  resourceError: string
  containers: string[]
  setResourceVersion: Dispatch<SetStateAction<string>>
}

export function getResourceParams() {
  const params = new URLSearchParams(decodeURIComponent(window.location.search))
  return {
    cluster: params.get('cluster') ?? '',
    kind: params.get('kind') ?? '',
    apiversion: params.get('apiversion') ?? '',
    namespace: params.get('namespace') ?? '',
    name: params.get('name') ?? '',
  }
}

export default function DetailsPage() {
  usePageVisitMetricHandler(Pages.searchDetails)
  const { t } = useTranslation()
  const navigate = useNavigate()
  const toast = useContext(AcmToastContext)
  const { settingsState } = useSharedAtoms()
  const vmActionsEnabled = useRecoilValue(settingsState)?.VIRTUAL_MACHINE_ACTIONS === 'enabled'
  const [resource, setResource] = useState<any>(undefined)
  const [containers, setContainers] = useState<string[]>()
  const [resourceVersion, setResourceVersion] = useState<string>('')
  const [resourceError, setResourceError] = useState('')
  const [resourceActionsOpen, setResourceActionsOpen] = useState(false)
  const [isDeleteResourceModalOpen, setIsDeleteResourceModalOpen] = useState(false)
  const { cluster, kind, apiversion, namespace, name } = getResourceParams()

  useEffect(() => {
    if (resourceVersion !== resource?.metadata.resourceVersion || name !== resource?.metadata.name) {
      /* istanbul ignore else */
      if (cluster === 'local-cluster') {
        getResource<IResource>({
          apiVersion: apiversion,
          kind,
          metadata: { namespace, name },
        })
          .promise.then((response) => {
            setResource(response)
            setResourceVersion(response?.metadata?.resourceVersion ?? '')
          })
          .catch((err) => {
            console.error('Error getting resource: ', err)
            setResourceError(err.message)
          })
      } else {
        fireManagedClusterView(cluster, kind, apiversion, name, namespace)
          .then((viewResponse) => {
            if (viewResponse?.message) {
              setResourceError(viewResponse.message)
            } else {
              setResource(viewResponse?.result)
              setResourceVersion(viewResponse?.result?.metadata.resourceVersion ?? '')
            }
          })
          .catch((err) => {
            console.error('Error getting resource: ', err)
            setResourceError(err)
          })
      }
    }
  }, [
    cluster,
    kind,
    apiversion,
    name,
    namespace,
    resourceVersion,
    resource?.metadata.resourceVersion,
    resource?.metadata.name,
  ])

  useEffect(() => {
    setContainers((resource && resource.spec?.containers?.map((container: any) => container.name)) ?? [])
  }, [resource])

  const location: {
    pathname: string
    state: {
      from: string
      search: string
      fromSearch: string
    }
  } = useLocation()

  const breadcrumbLink: string = useMemo(() => {
    const prevLocState = location.state
    if (prevLocState && prevLocState.from === NavigationPath.search) {
      // If we came to resources page from search - return to search with previous search filters
      const previousSearchState = location.state.fromSearch ?? /* istanbul ignore next */ ''
      return `${NavigationPath.search}${previousSearchState}`
    } else {
      /* istanbul ignore next */
      return NavigationPath.search
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  const searchDetailsContext = useMemo<SearchDetailsContext>(
    () => ({
      cluster,
      name,
      namespace,
      apiversion,
      kind,
      resource,
      resourceLoading: !resource && resourceError === '',
      resourceError,
      containers: containers || [],
      setResourceVersion,
    }),
    [apiversion, cluster, containers, kind, name, namespace, resource, resourceError]
  )

  const getResourceActions = useMemo(() => {
    const actions = [
      <DropdownItem
        component="button"
        key="edit-resource"
        onClick={() => {
          navigate(`${NavigationPath.resourceYAML}${window.location.search}`)
        }}
      >
        {t('Edit {{resourceKind}}', { resourceKind: kind })}
      </DropdownItem>,
      <DropdownItem
        component="button"
        key="delete-resource"
        onClick={() => {
          setIsDeleteResourceModalOpen(true)
        }}
      >
        {t('Delete {{resourceKind}}', { resourceKind: kind })}
      </DropdownItem>,
    ]
    if (vmActionsEnabled && kind.toLowerCase() === 'virtualmachine') {
      actions.unshift(
        ...[
          {
            action: 'Start',
            hubPath: `/apis/subresources.kubevirt.io/v1/namespaces/${namespace}/virtualmachines/${name}/start`,
            managedPath: '/virtualmachines/start',
          },
          {
            action: 'Stop',
            hubPath: `/apis/subresources.kubevirt.io/v1/namespaces/${namespace}/virtualmachines/${name}/stop`,
            managedPath: '/virtualmachines/stop',
          },
          {
            action: 'Restart',
            hubPath: `/apis/subresources.kubevirt.io/v1/namespaces/${namespace}/virtualmachines/${name}/restart`,
            managedPath: '/virtualmachines/restart',
          },
          {
            action: 'Pause',
            hubPath: `/apis/subresources.kubevirt.io/v1/namespaces/${namespace}/virtualmachineinstances/${name}/pause`,
            managedPath: '/virtualmachineinstances/pause',
          },
          {
            action: 'Unpause',
            hubPath: `/apis/subresources.kubevirt.io/v1/namespaces/${namespace}/virtualmachineinstances/${name}/unpause`,
            managedPath: '/virtualmachineinstances/unpause',
          },
        ].map((action) => (
          <DropdownItem
            key={`${action.action}-vm-resource`}
            component="button"
            onClick={() =>
              handleVMActions(
                action.action.toLowerCase(),
                cluster === 'local-cluster' ? action.hubPath : action.managedPath,
                { cluster, name, namespace },
                () => setResourceVersion(''), // trigger resource refetchto update details page data.
                toast,
                t
              )
            }
          >
            {t(`{{action}} {{resourceKind}}`, { action: action.action, resourceKind: kind })}
          </DropdownItem>
        )),
        <Divider key={'action-divider'} />
      )
    }
    return actions
  }, [cluster, kind, name, namespace, vmActionsEnabled, navigate, toast, t])

  return (
    <AcmPage
      header={
        <AcmPageHeader
          title={name}
          breadcrumb={[
            {
              text: t('Search'),
              to: breadcrumbLink,
            },
            {
              text: t('{{kind}} details', { kind }),
              to: '',
            },
          ]}
          navigation={
            <AcmSecondaryNav>
              <AcmSecondaryNavItem isActive={location.pathname === NavigationPath.resources}>
                <Link to={`${NavigationPath.resources}${window.location.search}`}>{t('Details')}</Link>
              </AcmSecondaryNavItem>
              <AcmSecondaryNavItem isActive={location.pathname === NavigationPath.resourceYAML}>
                <Link to={`${NavigationPath.resourceYAML}${window.location.search}`}>{t('YAML')}</Link>
              </AcmSecondaryNavItem>
              <AcmSecondaryNavItem isActive={location.pathname === NavigationPath.resourceRelated}>
                <Link to={`${NavigationPath.resourceRelated}${window.location.search}`}>{t('Related resources')}</Link>
              </AcmSecondaryNavItem>
              {(kind.toLowerCase() === 'pod' || kind.toLowerCase() === 'pods') && (
                <AcmSecondaryNavItem isActive={location.pathname === NavigationPath.resourceLogs}>
                  <Link to={`${NavigationPath.resourceLogs}${window.location.search}`}>{t('Logs')}</Link>
                </AcmSecondaryNavItem>
              )}
            </AcmSecondaryNav>
          }
          actions={
            <Dropdown
              isOpen={resourceActionsOpen}
              position={'right'}
              toggle={
                <DropdownToggle onToggle={() => setResourceActionsOpen(!resourceActionsOpen)}>
                  {t('Actions')}
                </DropdownToggle>
              }
              dropdownItems={getResourceActions}
              onSelect={() => setResourceActionsOpen(false)}
            />
          }
        />
      }
    >
      <DeleteResourceModal
        open={isDeleteResourceModalOpen}
        close={() => setIsDeleteResourceModalOpen(false)}
        resource={resource}
        cluster={cluster}
      />
      <Outlet context={searchDetailsContext} />
    </AcmPage>
  )
}

export function useSearchDetailsContext() {
  return useOutletContext<SearchDetailsContext>()
}
