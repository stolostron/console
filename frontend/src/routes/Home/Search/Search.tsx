/* Copyright Contributors to the Open Cluster Management project */
import { Redirect, Route, Switch } from 'react-router-dom'
import { NavigationPath } from '../../../NavigationPath'
import DetailsPage from './Details/DetailsPage'
import SearchPage from './SearchPage'

export default function Search() {
  return (
    <Switch>
      <Route exact path={NavigationPath.resources} component={DetailsPage} />
      <Route path={NavigationPath.resourceYAML} component={DetailsPage} />
      <Route path={NavigationPath.resourceLogs} component={DetailsPage} />
      <Route exact path={NavigationPath.search} component={SearchPage} />
      <Route path="*">
        <Redirect to={NavigationPath.search} />
      </Route>
    </Switch>
  )
}
