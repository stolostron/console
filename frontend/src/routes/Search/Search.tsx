/* Copyright Contributors to the Open Cluster Management project */
import { useEffect } from 'react'
import { Navigate, Route, Routes, useLocation, useNavigate } from 'react-router-dom-v5-compat'
import { createRoutePathFunction, NavigationPath } from '../../NavigationPath'
import { SearchAlertGroupProvider } from './components/SearchAlertGroup'
import DetailsOverviewPage from './Details/DetailsOverviewPage'
import DetailsPage from './Details/DetailsPage'
import LogsPage from './Details/LogsPage'
import RelatedResourceDetailsTab from './Details/RelatedResourceDetailsTab'
import YAMLPage from './Details/YAMLPage'
import SearchPage from './SearchPage'

const searchChildPath = createRoutePathFunction(NavigationPath.search)

export default function Search() {
  const location = useLocation()
  const navigate = useNavigate()

  useEffect(() => {
    // Redirect: 2.12 moves search & details pages from /multicloud/home/search -> /multicloud/search
    if (location.pathname.startsWith('/multicloud/home/search')) {
      navigate({
        pathname: location.pathname.replace('/home', ''),
        search: location.search,
      })
    }
  }, [location.pathname, location.search, navigate])

  return (
    <SearchAlertGroupProvider>
      <Routes>
        <Route element={<DetailsPage />}>
          <Route path={searchChildPath(NavigationPath.resourceYAML)} element={<YAMLPage />} />
          <Route path={searchChildPath(NavigationPath.resourceRelated)} element={<RelatedResourceDetailsTab />} />
          <Route path={searchChildPath(NavigationPath.resourceLogs)} element={<LogsPage />} />
          <Route path={searchChildPath(NavigationPath.resources)} element={<DetailsOverviewPage />} />
        </Route>
        <Route path={searchChildPath(NavigationPath.search)} element={<SearchPage />} />
        <Route path="*" element={<Navigate to={NavigationPath.search} replace />} />
      </Routes>
    </SearchAlertGroupProvider>
  )
}
