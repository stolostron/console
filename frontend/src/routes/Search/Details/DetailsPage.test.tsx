/* Copyright Contributors to the Open Cluster Management project */
// Copyright (c) 2021 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project

import { render, screen, waitFor } from '@testing-library/react'
import { MemoryRouter, Route, Routes } from 'react-router-dom-v5-compat'
import { RecoilRoot } from 'recoil'
import { nockGet, nockIgnoreApiPaths, nockIgnoreRBAC, nockPostRequest } from '../../../lib/nock-util'
import { waitForNocks } from '../../../lib/test-util'
import { NavigationPath } from '../../../NavigationPath'
import Search from '../Search'
import { getResourceParams } from './DetailsPage'

jest.mock('react-router-dom-v5-compat', () => {
  const originalModule = jest.requireActual('react-router-dom-v5-compat')
  return {
    __esModule: true,
    ...originalModule,
    useLocation: () => ({
      pathname: '/multicloud/search/resources',
      search: '?cluster=local-cluster&kind=Pod&apiversion=v1&namespace=testNamespace&name=testLocalPod',
      state: {
        from: '/multicloud/search',
        fromSearch: '?filters={%22textsearch%22:%22kind%3APod%22}',
      },
    }),
    useNavigate: () => jest.fn(),
  }
})
Object.defineProperty(window, 'location', {
  value: {
    pathname: '/multicloud/search/resources',
    search: '?cluster=local-cluster&kind=Pod&apiversion=v1&namespace=testNamespace&name=testLocalPod',
    state: {
      from: '/multicloud/search',
      fromSearch: '?filters={%22textsearch%22:%22kind%3APod%2',
    },
  },
})

const mockLocalClusterPod = {
  kind: 'Pod',
  apiVersion: 'v1',
  metadata: {
    name: 'testLocalPod',
    namespace: 'testNamespace',
    resourceVersion: '100702',
    creationTimestamp: '2022-11-18T11:33:56Z',
  },
  spec: {
    restartPolicy: 'Always',
    terminationGracePeriodSeconds: 30,
    dnsPolicy: 'ClusterFirst',
    nodeSelector: { 'kubernetes.io/os': 'linux' },
    containers: [
      {
        name: 'testContainer',
        resources: { requests: { cpu: '10m', memory: '50Mi' } },
        livenessProbe: {
          initialDelaySeconds: 10,
          timeoutSeconds: 5,
          periodSeconds: 10,
          successThreshold: 1,
          failureThreshold: 3,
        },
        readinessProbe: {
          initialDelaySeconds: 5,
          timeoutSeconds: 5,
          periodSeconds: 10,
          successThreshold: 1,
          failureThreshold: 3,
        },
      },
    ],
  },
  status: {
    phase: 'Running',
    conditions: [
      { type: 'Ready', status: 'True', lastProbeTime: null, lastTransitionTime: '2022-11-18T11:34:09Z' },
      {
        type: 'ContainersReady',
        status: 'True',
        lastProbeTime: null,
        lastTransitionTime: '2022-11-18T11:34:09Z',
      },
    ],
  },
}

describe('DetailsPage', () => {
  beforeEach(async () => {
    nockIgnoreRBAC()
    nockIgnoreApiPaths()
  })
  const metricNock = nockPostRequest('/metrics?search-details', {})

  it('should render local-cluster resource details correctly', async () => {
    render(
      <RecoilRoot>
        <MemoryRouter initialEntries={[NavigationPath.resources]}>
          <Routes>
            <Route path={`${NavigationPath.search}/*`} element={<Search />} />
          </Routes>
        </MemoryRouter>
      </RecoilRoot>
    )

    // Wait for delete resource requests to finish
    await waitForNocks([metricNock, nockGet(mockLocalClusterPod)])

    // Test that the component has rendered correctly with data
    await waitFor(() =>
      expect(
        screen.getByRole('heading', {
          name: /testlocalpod/i,
        })
      ).toBeTruthy()
    )

    // Wait for the details page to be loaded
    await waitFor(() =>
      expect(
        screen.getByRole('heading', {
          name: /pod details/i,
        })
      ).toBeTruthy()
    )
  })

  test('Should return the url search params correctly', () => {
    const res = getResourceParams()
    expect(res).toMatchSnapshot()
  })

  test('Should return the url search params incorrectly', () => {
    Object.defineProperty(window, 'location', {
      value: {
        pathname: '/multicloud/search/resources',
        search: '?',
        state: {
          from: '/multicloud/search',
          fromSearch: '?filters={%22textsearch%22:%22kind%3APod%2',
        },
      },
    })
    const res = getResourceParams()
    expect(res).toMatchSnapshot()
  })
})
