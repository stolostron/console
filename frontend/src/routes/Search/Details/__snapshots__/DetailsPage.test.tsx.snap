// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsPage Should return the url search params correctly 1`] = `
Object {
  "apiversion": "v1",
  "cluster": "local-cluster",
  "isHubClusterResource": true,
  "kind": "Pod",
  "name": "testLocalPod",
  "namespace": "testNamespace",
}
`;

exports[`DetailsPage Should return the url search params incorrectly 1`] = `
Object {
  "apiversion": "",
  "cluster": "",
  "isHubClusterResource": false,
  "kind": "",
  "name": "",
  "namespace": "",
}
`;
